### PR TITLE
MNT: Enable wheels for Python 3.12

### LINKED ISSUE #26557: [ENH]: Nightly Python 3.12 builds
### Problem

Trying to get pre-release testing on 3.12 working across some repositories I work on. Matplotlib is currently a blocker for projects that have it as a mandatory dependency. With numpy 1.26.0b1 on PyPI, I suspect this is now doable?

Looking through the issue/PR list, this isn't obviously on the roadmap, so I'm hoping this is a nudge, not pestering.

### Proposed solution

If it's just a matter of duplicating the following with `cp312`, I'm happy to submit a PR:

https://github.com/matplotlib/matplotlib/blob/0b5940f544c5715220a9f3d57c971d3fc4073b64/.github/workflows/cibuildwheel.yml#L138-L145

### PR DESCRIPTION
**Pull request by QuLogic**
## PR summary

We should probably figure out tests #24711, but that shouldn't stop us from making the wheels at least.

Fixes #26557

## PR checklist
- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by effigies** [comment]
FWIW, I have had some success [using the `requires` table](https://github.com/nipy/nitime/blob/8a01879b07d078ea11e592c5ab07b8d88426ef3d/pyproject.toml#L1-L12) to pin specific versions for building:

```TOML
[build-system]
requires = [
  "setuptools",
  "setuptools_scm[toml]>=6.2",
  "cython",
  # Newer than NEP29-minimum: compile against oldest numpy available
  "numpy==1.26b1; python_version >= '3.12rc1'",
  "numpy==1.24; python_version >= '3.11' and python_version < '3.12.dev0'",
  # NEP29-minimum as of Aug 17, 2023
  "numpy==1.22; python_version >= '3.8' and python_version < '3.11'",
]
build-backend = "setuptools.build_meta"
```

**Comment by QuLogic** [comment]
> FWIW, I have had some success [using the `requires` table](https://github.com/nipy/nitime/blob/8a01879b07d078ea11e592c5ab07b8d88426ef3d/pyproject.toml#L1-L12) to pin specific versions for building:

Yes, I did see that elsewhere, though I was hoping to avoid it since it's mostly a wheel-building issue.

> ```toml
>   # Newer than NEP29-minimum: compile against oldest numpy available
>   "numpy==1.26b1; python_version >= '3.12rc1'",
>   "numpy==1.24; python_version >= '3.11' and python_version < '3.12.dev0'",
>   # NEP29-minimum as of Aug 17, 2023
>   "numpy==1.22; python_version >= '3.8' and python_version < '3.11'",
> ```

Note though that since [NumPy 1.25](https://numpy.org/doc/stable/release/1.25.0-notes.html#compiling-against-the-numpy-c-api-is-now-backwards-compatible-by-default), we no longer use NEP29 as the basis for the _build_ requirement.

**Comment by effigies** [comment]
> Note though that since [NumPy 1.25](https://numpy.org/doc/stable/release/1.25.0-notes.html#compiling-against-the-numpy-c-api-is-now-backwards-compatible-by-default), we no longer use NEP29 as the basis for the _build_ requirement.

I just saw that (via #26443), and look forward to removing the extra lines from my projects.

**Comment by QuLogic** [comment]
Ah, no wheels for contourpy; I really don't want to have to figure out how to get full Meson support in there too; so waiting on https://github.com/contourpy/contourpy/issues/262

File: .github/workflows/cibuildwheel.yml [hunk 1]
```diff
@@ -135,6 +135,24 @@ jobs:
...
+      - name: Build wheels for CPython 3.12
+        uses: pypa/cibuildwheel@39a63b5912f086dd459cf6fcb13dcdd3fe3bc24d # v2.15.0
+        with:
+          package-dir: dist/${{ needs.build_sdist.outputs.SDIST_NAME }}
+        env:
+          CIBW_BUILD: "cp312-*"
+          CIBW_ARCHS: ${{ matrix.cibw_archs }}
+          # Remove this once NumPy with Python 3.12 wheels is not pre-release.
+          CIBW_BEFORE_BUILD: >-
+            pip install certifi "pybind11>=2.6" "setuptools>=42" "setuptools_scm>=7" &&
```
Comment (effigies) [inline .github/workflows/cibuildwheel.yml:147]: Would you be willing to try adding contourpy's [build-requires](https://github.com/contourpy/contourpy/blob/c963ecc77ec3ab12b7259cf0fef8c1205479adf3/pyproject.toml#L4-L6)? They're not likely to put out a wheel until numpy puts out a final 1.26.0 release, which is not going to happen until 3.12 final.

```suggestion
            pip install certifi "setuptools>=42" "setuptools_scm>=7" "meson >= 1.2.0" "meson-python >= 0.13.1" "pybind11 >= 2.10.4"
```

Comment (neutrinoceros) [reply .github/workflows/cibuildwheel.yml]: Where are you getting this from ? CPython 3.12 .0 is scheduled for early October while numpy 1.26 's milestone is currently aimed at September 1st.

Comment (effigies) [reply .github/workflows/cibuildwheel.yml:147]: Possibly a faulty memory. Apologies.

Comment (neutrinoceros) [reply .github/workflows/cibuildwheel.yml:147]: no problem, just asking :)

**Review by oscargus — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Comment by ksunden** [comment]
The backport was included in #26672

**Pull request merged**